src/test/os/bluestore/TestBlueStoreSlowFastCoDel.cc

```diff
@@ -0,0 +1,191 @@
+#include <mutex>
+#include <cmath>
+#include <vector>
+#include <condition_variable>
+#include <cmath>
+#include <cstdlib>
+
+#include "gtest/gtest.h"
+#include "include/Context.h"
+
+#include "common/ceph_time.h"
+#include "os/bluestore/BlueStoreSlowFastCoDel.h"
+
+
+static int64_t milliseconds_to_nanoseconds(int64_t ms) {
+  return ms * 1000.0 * 1000.0;
+}
+
+static double nanoseconds_to_milliseconds(int64_t ms) {
+  return ms / (1000.0 * 1000.0);
+}
+
+class BlueStoreSlowFastCoDelMock : public BlueStoreSlowFastCoDel {
+public:
+  BlueStoreSlowFastCoDelMock(
+    CephContext *_cct,
+    std::function<void(int64_t)> _bluestore_budget_reset_callback,
+    std::function<int64_t()> _get_kv_throttle_current,
+    std::mutex &_iteration_mutex,
+    std::condition_variable &_iteration_cond,
+    int64_t _target_latency,
+    int64_t _fast_interval,
+    int64_t _slow_interval,
+    double _target_slope
+  ) : BlueStoreSlowFastCoDel(_cct, _bluestore_budget_reset_callback,
+                             _get_kv_throttle_current),
+      iteration_mutex(_iteration_mutex), iteration_cond(_iteration_cond),
+      test_target_latency(_target_latency), test_fast_interval(_fast_interval),
+      test_slow_interval(_slow_interval), test_target_slope(_target_slope) {
+    init_test();
+  }
+
+  void init_test() {
+    std::lock_guard l(register_lock);
+    activated = true;
+    target_slope = test_target_slope;
+    slow_interval = test_slow_interval;
+    initial_fast_interval = test_fast_interval;
+    min_target_latency = milliseconds_to_nanoseconds(1);
+    initial_target_latency = test_target_latency;
+    max_target_latency = milliseconds_to_nanoseconds(500);
+    initial_bluestore_budget = 100 * 1024;
+    min_bluestore_budget = 10 * 1024;
+    bluestore_budget_increment = 1024;
+    regression_history_size = 5;
+    bluestore_budget = initial_bluestore_budget;
+    min_bluestore_budget = initial_bluestore_budget;
+    max_queue_length = min_bluestore_budget;
+    fast_interval = initial_fast_interval;
+    target_latency = initial_target_latency;
+    min_latency = INITIAL_LATENCY_VALUE;
+    slow_interval_registered_bytes = 0;
+    regression_throughput_history.clear();
+    regression_target_latency_history.clear();
+    slow_interval_start = ceph::mono_clock::zero();
+  }
+
+  std::vector <int64_t> target_latency_vector;
+
+protected:
+  std::mutex &iteration_mutex;
+  std::condition_variable &iteration_cond;
+  int64_t test_target_latency;
+  int64_t test_fast_interval;
+  int64_t test_slow_interval;
+  double test_target_slope;
+
+  void on_fast_interval_finished() override {
+    std::unique_lock <std::mutex> locker(iteration_mutex);
+    iteration_cond.notify_one();
+  }
+
+  void on_slow_interval_finished() override {
+    target_latency_vector.push_back(target_latency);
+  }
+};
+
+class TestSlowFastCoDel : public ::testing::Test {
+public:
+  CephContext *ceph_context = nullptr;
+  BlueStoreSlowFastCoDelMock *slow_fast_codel = nullptr;
```

Comment:
std::unique_ptr -- unless you have a really good reason, we really don't allow bare pointers like this.